Allow to fingerprint Blazor.js

maraf opened the pull request:
This PR allows to put hard fingerprint on blazor.js file if `WriteImportMapToHtml=true`. 
It can additionally be controlled with `BlazorFingerprintBlazorJs=false`.

maraf committed:
Allow to fingerprint Blazor.js

maraf committed:
Merge remote-tracking branch 'upstream/main' into BlazorJsFingerprint

maraf committed:
Fix relative path for publish fingerprinted asset

maraf committed:
Update test to assert fingerprinted blazor.js

maraf commented:
There is a bit of dance to make asset promotion in publish happy. The implementation is based on what we do in runtime to promote hard fingerprinted assets for publish. An alternative would probably be to define these assets as build only and create new assets for publish. I'm not sure if that's a better solution and whether the longer-term goal of not copying these assets to build outputs isn't a way to go

maraf committed:
Fix testasset used by other tests

maraf commented:
/azp list

maraf commented:
/azp run list

javiercn commented on src/BlazorWasmSdk/Targets/Microsoft.NET.Sdk.BlazorWebAssembly.6_0.targets:
```diff
@@ -143,6 +156,61 @@ Copyright (c) .NET Foundation. All rights reserved.
+      AssetKind="All"
+      AssetMergeSource="$(StaticWebAssetMergeTarget)"
+      AssetRole="Primary"
+      AssetTraitName="WasmResource"
+      AssetTraitValue="boot"
+      CopyToOutputDirectory="Never"
+      CopyToPublishDirectory="PreserveNewest"
+      ContentRoot="%(_BlazorJSStaticWebAsset.ContentRoot)"
+      BasePath="%(_BlazorJSStaticWebAsset.BasePath)"
+    >
+      <Output TaskParameter="Assets" ItemName="_BlazorJSPublishStaticWebAssets" />
+    </DefineStaticWebAssets>
+    <DefineStaticWebAssetEndpoints
+      CandidateAssets="@(_BlazorJSPublishStaticWebAssets)"
+      ExistingEndpoints="@(StaticWebAssetEndpoint)"
+      ContentTypeMappings="@(StaticWebAssetContentTypeMapping)"
+    >
+      <Output TaskParameter="Endpoints" ItemName="_BlazorJSPublishStaticWebAssetsEndpoint" />
+    </DefineStaticWebAssetEndpoints>
+    <PropertyGroup>
+      <_BlazorJSStaticWebAssetFullPath>@(_BlazorJSStaticWebAsset->'%(FullPath)')</_BlazorJSStaticWebAssetFullPath>
+    </PropertyGroup>
+    <ItemGroup>
+      <_BlazorJSStaticWebAsset Include="@(StaticWebAsset)" Condition="'%(AssetTraitName)' == 'Content-Encoding' and '%(RelatedAsset)' == '$(_BlazorJSStaticWebAssetFullPath)'" />
+    </ItemGroup>
+    <FilterStaticWebAssetEndpoints Condition="'@(_BlazorJSStaticWebAsset)' != ''"
+      Endpoints="@(StaticWebAssetEndpoint)"
+      Assets="@(_BlazorJSStaticWebAsset)"
+      Filters=""
+    >
+      <Output TaskParameter="FilteredEndpoints" ItemName="_BlazorJSEndpointsToRemove" />
+    </FilterStaticWebAssetEndpoints>
+    <ItemGroup>
+      <StaticWebAsset Remove="@(_BlazorJSStaticWebAsset)" />
+      <StaticWebAsset Include="@(_BlazorJSPublishStaticWebAssets)" />
+      <StaticWebAssetEndpoint Remove="@(_BlazorJSEndpointsToRemove)" />
+      <StaticWebAssetEndpoint Include="@(_BlazorJSPublishStaticWebAssetsEndpoint)" />
+    </ItemGroup>
+  </Target>
+
```
This part is confusing me a bit, do we rewrite blazor.js in any way during this whole process? If not, I'm not sure why this is needed at all.

maraf replied:
It's the same issue we have with runtime assets. If we do hard fingerprint during build, the asset to promote for publish has identity without fingerprint (`bin/Debug/wwwroot/_framework/blazor.js`), but the file on disk has fingerprint (`bin/Debug/wwwroot/_framework/blazor.abcd.js`) and thus it complains that the file doesn't exist.

This is fixing by creating asset that has fingerprint in identity.

Maybe we can do soft fingerprint during build and hard only during publish. Any ideas?

maraf committed:
Merge branch 'main' into BlazorJsFingerprint

maraf committed:
Merge remote-tracking branch 'upstream/main' into BlazorJsFingerprint

maraf committed:
Merge remote-tracking branch 'upstream/main' into BlazorJsFingerprint

maraf committed:
Use BlazorFingerprintBlazorJs to switch FingerprintCandidates instead of defining patterns

Aligns with WasmFingerprintAssets and when false it doesn't produce a hard nor soft fingerprint
